boa/src/syntax/lexer/cursor.rs

```diff
@@ -0,0 +1,286 @@
+                ));
+            }
+        }
+    }
+
+    /// Fills the buffer with characters until the first character (x) for which the predicate (pred) is false
+    /// (or the next character is none).
+    ///
+    /// Note that all characters up until x are added to the buffer including the character right before.
+    pub(super) fn take_while_pred<F>(&mut self, buf: &mut String, pred: &F) -> io::Result<()>
+    where
+        F: Fn(char) -> bool,
+    {
+        let _timer = BoaProfiler::global().start_event("cursor::take_while_pred()", "Lexing");
+
+        loop {
+            if !self.next_is_pred(pred)? {
+                return Ok(());
+            } else if let Some(ch) = self.next_char()? {
+                buf.push(ch);
+            } else {
+                // next_is_pred will return false if the next value is None so the None case should already be handled.
+                unreachable!();
+            }
+        }
+    }
+
+    /// It will fill the buffer with checked ASCII bytes.
+    ///
+    /// This expects for the buffer to be fully filled. If it's not, it will fail with an
+    /// `UnexpectedEof` I/O error.
+    #[inline]
+    pub(super) fn fill_bytes(&mut self, buf: &mut [u8]) -> io::Result<()> {
+        let _timer = BoaProfiler::global().start_event("cursor::fill_bytes()", "Lexing");
+
+        self.iter.fill_bytes(buf)
+    }
+
+    /// Retrieves the next UTF-8 character.
+    #[inline]
+    pub(crate) fn next_char(&mut self) -> Result<Option<char>, Error> {
+        let _timer = BoaProfiler::global().start_event("cursor::next_char()", "Lexing");
+
+        let chr = match self.peeked.take() {
+            Some(v) => v,
+            None => self.iter.next_char()?,
+        };
+
+        match chr {
+            Some('\r') => self.carriage_return(),
+            Some('\n') | Some('\u{2028}') | Some('\u{2029}') => self.next_line(),
+            Some(_) => self.next_column(),
+            None => {}
+        }
+
+        Ok(chr)
+    }
+}
+
+/// Inner iterator for a cursor.
+#[derive(Debug)]
+struct InnerIter<R> {
+    iter: Bytes<R>,
+}
+
+impl<R> InnerIter<R> {
+    /// Creates a new inner iterator.
+    #[inline]
+    fn new(iter: Bytes<R>) -> Self {
+        Self { iter }
+    }
+}
+
+impl<R> InnerIter<R>
+where
+    R: Read,
+{
+    /// It will fill the buffer with checked ASCII bytes.
+    ///
+    /// This expects for the buffer to be fully filled. If it's not, it will fail with an
+    /// `UnexpectedEof` I/O error.
+    #[inline]
+    fn fill_bytes(&mut self, buf: &mut [u8]) -> io::Result<()> {
+        for byte in buf.iter_mut() {
+            *byte = self.next_ascii()?.ok_or_else(|| {
+                io::Error::new(
+                    io::ErrorKind::UnexpectedEof,
+                    "unexpected EOF when filling buffer",
+                )
+            })?;
+        }
+        Ok(())
+    }
+
+    /// Retrieves the next UTF-8 checked character.
+    #[inline]
```

Comment:
I wouldn't inline this by default. It's a pretty big function, I would leave LLVM to do the heuristics, what do you think?
```suggestion
```

Reply:
Yeah makes sense